### PR TITLE
Volume operations and flows

### PR DESCRIPTION
**Pull request by shtripat**
@r0h4n @nthomas-redhat @nnDarshan please review this patch as its updated and works fine.

### CODE REVIEW

File: tendrl/gluster_bridge/atoms/volume/create.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 
 class Create(object):
     def start(self, name, bricks):
-        subprocess.call(['gluster', 'volume', 'create',
-                         name, ' '.join(bricks), ' force'])
+        cmd = ['gluster', 'volume', 'create', name]
+        cmd.extend(bricks)
+        cmd.append('force')
+        subprocess.call(cmd)
```
Comment (TimothyAsir) [inline tendrl/gluster_bridge/atoms/volume/create.py]: Why gluster volume start function contain has create functionality?

File: tendrl/gluster_bridge/manager/rpc.py [hunk 2]
```diff
@@ -45,6 +45,7 @@ def _acceptor(self):
                         break
                     except Exception as ex:
                         LOG.error(ex)
+            gevent.sleep(2)
```
Comment (TimothyAsir) [inline tendrl/gluster_bridge/manager/rpc.py]: Why sleep here!. May be you can give a comment note here.

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 3]
```diff
@@ -11,81 +11,432 @@ object_details:
     attrs:
       volname:
         type: String
+	value: etcd
+	help: Gluster volume name
```
Comment (TimothyAsir) [inline etc/sds_operations_gluster-3.8.3.yaml]: Remove the tabs. \t are not allowed for yaml

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 4]
```diff
@@ -101,48 +452,228 @@ object_details:
 
   peer:
     attrs:
-    - hostname
+      hostname
```
Comment (TimothyAsir) [inline etc/sds_operations_gluster-3.8.3.yaml]: ":" is missing for hostname

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 5]
```diff
@@ -101,48 +452,228 @@ object_details:
...
+	run: "tendrl.gluster_bridge.atoms.snapshot.config"
         type: "update"
+	conditions:
+	  pre_condition: None
+	  post_condition: None
+	name: "config_snapshot"
+	outputs:
+	  - "success"
+	global_inputs: None
+	local_inputs
```
Comment (TimothyAsir) [inline etc/sds_operations_gluster-3.8.3.yaml]: ':' is missing

File: tendrl/gluster_bridge/atoms/volume/create.py (hunk 1 above)
Comment (shtripat) [inline tendrl/gluster_bridge/atoms/volume/create.py]: This atom is for create volume only. But starts the volume after creation.

File: etc/sds_operations_gluster-3.8.3.yaml (hunk 3 above)
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: Will do that.

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 6]
```diff
@@ -11,81 +11,432 @@ object_details:
...
-
-          global_inputs:
-          - An immutable input_name => value dictionary which specifies any initial inputs that should be automatically injected into the atoms scope before the atom execution commences (this allows for providing atom local values that do not need to be provided by other atoms/dependents).
+            - "volname"
+          global_inputs: None
           local_inputs:
-          - A set or list of required inputs for this atom’s execute method
-          version: An immutable version that associates version information with this atom. It can be useful in resuming older versions of atoms. Standard major, minor versioning concepts should apply
+            - "volname"
+            - ["stripe_count"]
```
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: Actually wanted to add an optional parameter. Not sure how to ?

File: etc/sds_operations_gluster-3.8.3.yaml (hunk 4 above)
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: Will do that

File: etc/sds_operations_gluster-3.8.3.yaml (hunk 5 above)
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: Will add

**Comment by shtripat** [comment]
@TimothyAsir @nnDarshan @nthomas-redhat please review the patches. Incorporated the comments from Tim.

**Comment by shtripat** [comment]
@aravindavk @atinmu please review.

File: tendrl/gluster_bridge/flows/create_volume.py [hunk 7]
```diff
@@ -11,10 +11,50 @@ def __init__(self, api_job):
         self.atom = Create
 
     def start(self):
-        attributes = json.loads(self.api_job['attributes'].decode('utf-8'))
+        attributes = self.api_job['attributes']
+        print attributes
```
Comment (nnDarshan) [inline tendrl/gluster_bridge/flows/create_volume.py]: please remove debug messages in this file.

File: tendrl/gluster_bridge/flows/create_volume.py [hunk 8]
```diff
@@ -11,10 +11,50 @@ def __init__(self, api_job):
...
+        elif v_replica_count is not None:
+            print v_replica_count
+            self.atom().start(
+                vol_name,
+                brickdetails,
+                transport = v_transport,
+                replica_count = v_replica_count,
+                arbiter_count = v_arbiter_count
+            )
+            print "Hell"
```
Comment (nnDarshan) [inline tendrl/gluster_bridge/flows/create_volume.py]: was it frustrating :D

File: tendrl/gluster_bridge/flows/add_volume_brick.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import json
+
+import etcd
+from tendrl.gluster_bridge.atoms.volume.add_brick import AddBrick
+
+
+class AddVolumeBrick(object):
+    def __init__(self, api_job):
+        super(AddVolumeBrick, self).__init__()
```
Comment (nnDarshan) [inline tendrl/gluster_bridge/flows/add_volume_brick.py]: not needed I guess. (in all the constructors of flows)

File: tendrl/gluster_bridge/flows/add_volume_brick.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        self.atom = AddBrick
+
+    def start(self):
+        attributes = self.api_job['attributes']
+        vol_name = attributes['volname']
+        bricks = attributes['brickdetails']
+        self.atom().start(vol_name, bricks)
+        self.api_job['status'] = "finished"
+        etcd.Client().write(self.api_job['request_id'],
+                            json.dumps(self.api_job))
```
Comment (nnDarshan) [inline tendrl/gluster_bridge/flows/add_volume_brick.py]: I feel you can have the above 2 lines in rpc.py, in such a way that if flow succeeds you can have status as Finished else status as Failed. Thus you avoid having this in every flow file.

File: tendrl/gluster_bridge/atoms/peer/detach.py [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import subprocess
+
+
+class Detach(object):
+    def start(self, peer):
+        subprocess.call(['gluster', 'peer', 'detach',
```
Comment (shtripat) [inline tendrl/gluster_bridge/atoms/peer/detach.py]: I think this we would take up in a separate PR and yes for sure its there back of my mind. Thanks for suggestion :)

File: tendrl/gluster_bridge/atoms/volume/create.py [hunk 12]
```diff
@@ -2,7 +2,44 @@
...
+        replica_count = None,
+        arbiter_count = None,
+        stripe_count = None,
+        disperse_count = None,
+        disperse_data_count = None,
+        redundancy_count = None,
+        transport = [],
+        ):
+        cmd = ['gluster', 'volume', 'create', name]
+        print replica_count
```
Comment (shtripat) [inline tendrl/gluster_bridge/atoms/volume/create.py]: Ah somehow forgot to remove. Done.

File: tendrl/gluster_bridge/atoms/volume/create.py [hunk 13]
```diff
@@ -2,7 +2,44 @@
...
+                cmd.append('disperse_data')
+                cmd.append(disperse_data_count)
+        if transport:
+            cmd.append('transport')
+            cmd.append(','.join(transport))
+        cmd.extend(bricks)
+        cmd.append('force')
+        cmd.append('--mode=script')
+        subprocess.call(cmd)
+        subprocess.call(['gluster', 'volume', 'start',
```
Comment (shtripat) [inline tendrl/gluster_bridge/atoms/volume/create.py]: Thats good idea. @r0h4n @nthomas-redhat what you say? I am +1 for this.

File: tendrl/gluster_bridge/atoms/volume/delete.py [hunk 14]
```diff
@@ -0,0 +1,9 @@
+import subprocess
+
+
+class Delete(object):
+    def start(self, name):
+        subprocess.call(['gluster', 'volume', 'stop',
+                        name, '--mode=script'])
+        subprocess.call(['gluster', 'volume', 'delete',
```
Comment (shtripat) [inline tendrl/gluster_bridge/atoms/volume/delete.py]: +1

File: tendrl/gluster_bridge/flows/create_volume.py (hunk 8 above)
Comment (shtripat) [inline tendrl/gluster_bridge/flows/create_volume.py]: Haha. yeah will remove this ;)

File: tendrl/gluster_bridge/flows/add_volume_brick.py (hunk 9 above)
Comment (shtripat) [inline tendrl/gluster_bridge/flows/add_volume_brick.py]: oh ok. will try that. just retained as was there in original code of create volume flow.

File: tendrl/gluster_bridge/flows/add_volume_brick.py (hunk 10 above)
Comment (shtripat) [inline tendrl/gluster_bridge/flows/add_volume_brick.py]: good suggestion. need to try in another PR may be with return value check as well.

**Comment by shtripat** [comment]
@brainfunked @r0h4n @nthomas-redhat @anmolbabu @nnDarshan @TimothyAsir I have incorporated the changes guidelines for spec yaml file. Kindly review.

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 15]
```diff
@@ -1,160 +1,543 @@
...
+        outputs:
+          - "success"
+        inputs:
+          - "peer"
+        version: "0.1.0"
+
+flows:
+  CreateVolume:
+    enabled: True
+    tendrl_context: "$cluster.id"
```
Comment (r0h4n) [inline etc/sds_operations_gluster-3.8.3.yaml]: this needs more eyes, IMO this should be a list instead of single item

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 16]
```diff
@@ -1,160 +1,543 @@
...
+        help: "If force execute the action"
 
     atoms:
-        info:
-          type: "get"
         create:
-          run: "gluster_bridge.atoms.volume.create"
+          enabled: True
+          uuid: "242f6190-9b37-11e6-950d-a24fc0d9649c"
+          flows: ["CreateVolume"]
```
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: Do you to have flows like CreateDistributeVolume, CreateStripedVolume, CreateStripedReplicateVolume etc as separate flows?
Based on non-mandatory fields possible for gluster volumes this permutation is going to be lot big in no I feel. Dont you feel we should some mechanism to mention optional arguments and only if they are passed we should invoke flow with different arguments?

@brainfunked @nthomas-redhat @nnDarshan @anmolbabu thoughts?
I strongly feel we should have some mechanism to mention a optional list of arguments as well for flows and this is the right juncture when we are defining the structure of yaml.

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 17]
```diff
@@ -1,160 +1,543 @@
...
-
-          global_inputs:
-          - An immutable input_name => value dictionary which specifies any initial inputs that should be automatically injected into the atoms scope before the atom execution commences (this allows for providing atom local values that do not need to be provided by other atoms/dependents).
-          local_inputs:
-          - A set or list of required inputs for this atom’s execute method
-          version: An immutable version that associates version information with this atom. It can be useful in resuming older versions of atoms. Standard major, minor versioning concepts should apply
+            - "volname"
+          inputs:
+            - "volname"
+            - ["stripe_count"]
```
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: I didnt have a way to mention an optional argument so used square brackets (I feel it cannot be as square bracket means list here).
Now again my question is to have some way to mention optional arguments for the flows else we might end writing so many flows just because of optional arguments and their combinations in gluster.
I remember VDSM doing the same with optional arguments and we are doing something similar here.

Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: @r0h4n I have another question regarding listing of fields under `inputs` and `outputs`. Do we just need listing as its done in this patch or we need something like 

input: ['volname', 'brickdetails', 'force']

Kindly confirm with one example for `inputs` and `outputs` here may be ..

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 18]
```diff
@@ -1,160 +1,543 @@
...
+          enabled: True
+          uuid: "242f6190-9b37-11e6-950d-a24fc0d9649c"
+          flows: ["CreateVolume"]
+          run: "tendrl.gluster_bridge.atoms.volume.create.Create"
           type: "create"
-          conditions:
-            pre_condition : ["atoms.<atom_name>.<condition_function>1"...]
-            post_condition: ["atoms.<atom_name>.<condition_function>1"...]
-          name: Meaningful name for this atom, should be something that is distinguishable and understandable for notification, debugging, storing and any other similar purposes
+          pre_run : []
```
Comment (r0h4n) [inline etc/sds_operations_gluster-3.8.3.yaml]: Yes, and these functions will receive the "inputs" for that atom for validation of those "inputs"

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 19]
```diff
@@ -1,160 +1,543 @@
...
+          post_run: []
```
Comment (r0h4n) [inline etc/sds_operations_gluster-3.8.3.yaml]: Yes, we can have things done in post_run, but these actions should be atomic (reversible by the same atom)

File: etc/sds_operations_gluster-3.8.3.yaml (hunk 17 above)
Comment (r0h4n) [inline etc/sds_operations_gluster-3.8.3.yaml]: Its just a list for "inputs" and for "output" you need to use the structure described here 
https://github.com/Tendrl/bridge_common/blob/master/etc/samples/tendrl_definitions_generic-0.1.sample.yaml#L58

File: etc/sds_operations_gluster-3.8.3.yaml [hunk 20]
```diff
@@ -1,160 +1,543 @@
...
+          outputs:
+            - "success"
+          inputs:
+            - "volname"
+          version: "0.1.0"
   brick:
+    enabled: True
+    value: "/clusters/$cluster.id/volumes/$volume.id/bricks/$brick.id"
     attrs:
       brickdetails:
```
Comment (r0h4n) [inline etc/sds_operations_gluster-3.8.3.yaml]: then it is ok

File: etc/sds_operations_gluster-3.8.3.yaml (hunk 18 above)
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: Do you mean all the input fields would be passed for all the validate functions? It doesnt seem too good to me, but we can live with it I feel.

File: etc/sds_operations_gluster-3.8.3.yaml (hunk 19 above)
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: I have sent another patch post these comments with few validations added. Can you please check once and see if looks better?

File: etc/sds_operations_gluster-3.8.3.yaml (hunk 17 above)
Comment (shtripat) [inline etc/sds_operations_gluster-3.8.3.yaml]: Is it something like

outputs: [
  volname: {required: True},
  brickdetails: {required: True},
  replica_count: {required: False}
]

inputs: [volname, brickdetails, replica_count, ...]

For this reason only I wanted a working sample here :) Anyway plz confirm the same.

**Comment by shtripat** [comment]
@r0h4n @nthomas-redhat @brainfunked @anmolbabu @nnDarshan I have pushed fresh patch with latest input / output guidelines for atoms and flows with dummy validations added for volume/peer atoms.

Kindly review..

**Comment by r0h4n** [comment]
Fix "tox -epep8" failures https://travis-ci.org/Tendrl/gluster_bridge/builds/171148366

**Comment by shtripat** [comment]
@brainfunked @nthomas-redhat waiting for your reviews. please do the same.

**Comment by r0h4n** [comment]
Is this PR still required @shtripat

**Comment by r0h4n** [comment]
@shtripat Is this still required? Please attach spec or bug

**Comment by shtripat** [comment]
> @shtripat Is this still required? Please attach spec or bug

@r0h4n I would close this as its pretty old one and lot of changes have happened down the line.
